app/src/main/java/io/apicurio/registry/mt/TenantReaper.java

```diff
@@ -0,0 +1,148 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.apicurio.registry.mt;
+
+import io.apicurio.multitenant.api.beans.RegistryTenantList;
+import io.apicurio.multitenant.api.beans.SortBy;
+import io.apicurio.multitenant.api.beans.SortOrder;
+import io.apicurio.multitenant.api.beans.TenantStatusValue;
+import io.apicurio.multitenant.api.datamodel.RegistryTenant;
+import io.apicurio.multitenant.client.TenantManagerClient;
+import io.apicurio.registry.storage.RegistryStorage;
+import io.apicurio.registry.types.Current;
+import io.apicurio.registry.utils.OptionalBean;
+import io.quarkus.scheduler.Scheduled;
+import org.slf4j.Logger;
+
+import javax.annotation.PostConstruct;
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.List;
+import java.util.Random;
+
+import static io.quarkus.scheduler.Scheduled.ConcurrentExecution.SKIP;
+
+/**
+ * Periodically cleanup data of tenants marked as deleted.
+ *
+ * @author Jakub Senko <jsenko@redhat.com>
+ */
+@ApplicationScoped
+public class TenantReaper {
+
+    @Inject
+    Logger log;
+
+    @Inject
+    MultitenancyProperties properties;
+
+    @Inject
+    TenantMetadataService tenantService;
+
+    @Inject
+    @Current
+    RegistryStorage storage;
+
+    @Inject
+    TenantContextLoader tcl;
+
+    @Inject
+    TenantContext tctx;
+
+    @Inject
+    OptionalBean<TenantManagerClient> tenantManagerClient;
+
+    Instant next;
+
+    @PostConstruct
+    void init() {
+        if (!properties.isMultitenancyEnabled()) {
+            return;
+        }
+        int stagger = 0;
+        // Only stagger if the reaper period is at least 1 minute (testing support).
+        if (properties.getReaperPeriod().compareTo(Duration.ofSeconds(60)) >= 0) {
+            // Start with a random stagger, 1-30 minutes, inclusive.
+            stagger = new Random().nextInt(30) + 1;
+            log.debug("Staggering tenant reaper job by {} minutes", stagger);
+        }
+        next = Instant.now().plus(Duration.ofSeconds(stagger * 60L));
+    }
+
+    /**
+     * Minimal granularity is 1 minute.
+     */
+    @Scheduled(concurrentExecution = SKIP, every = "{registry.multitenancy.reaper.every}")
+    void run() {
+        if (!properties.isMultitenancyEnabled()) {
+            return;
+        }
+        final Instant now = Instant.now();
+        if (now.isAfter(next)) {
+            try {
+                log.debug("Running tenant reaper job at {}", now);
+                reap();
+                // Only force cache invalidation if the reaper period is less than 1 minute (testing support).
+                if (properties.getReaperPeriod().compareTo(Duration.ofSeconds(60)) < 0) {
```

Comment:
why is this complete cache invalidation needed? I think we don't need it, and actually having this invalidation forced here could hide some bugs in the future.

Reply:
It's because of the integration tests:
1. After tenants are deleted, their deleted status is cached in the loader.
2. The API to access those tenants is disabled.
3. To test that the data has been actually deleted, I force the status back to READY in the tenant manager. However, the cache prevents these changes to take effect immediately. In normal operation, this would not happen, or the cache would eventually refresh. But unlike deletion, reaper does not execute when the status is READY so there's nowhere to do the refresh. Therefore, either we create a property that will control cache timeout, or it has to be done this way. I'm open to doing it using an additional property.

Reply:
I can make the stagger configurable as well, may depend on the # of nodes

Reply:
Or if we have a more generalized property that hints to the registry the # of nodes, that would may be even better solution. Could be used in other places.

Reply:
ok makes sense, so in this case it's ok to leave this as is for now, thank you!